core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -546,6 +574,103 @@ impl<Block: BlockT> Inner<Block> {
 		Action::ProcessAndDiscard(topic, benefit::BASIC_VALIDATED_COMMIT)
 	}
 
+	fn validate_catch_up_message(&mut self, who: &PeerId, full: &FullCatchUpMessage<Block>)
+		-> Action<Block::Hash>
+	{
+		// FIXME: adjust costs
+		match self.pending_catch_up_request.take() {
+			Some((peer, request, instant)) => {
+				if peer != *who {
+					self.pending_catch_up_request = Some((peer, request, instant));
+					return Action::Discard(-1);
+				}
+
+				if request.set_id != full.set_id {
+					return Action::Discard(-1);
+				}
+
+				if request.round.0 > full.message.round_number {
+					return Action::Discard(-1);
+				}
+
+				// move request to pending import state we won't push out any
+				// catch up requests until we import this one (either with a
+				// success or failure)
+				self.pending_catch_up_import = Some((peer, request, instant));
+
+				let topic = super::global_topic::<Block>(full.set_id.0);
+				Action::ProcessAndDiscard(topic, 0)
+			},
+			None => Action::Discard(-1),
+		}
+	}
+
+	fn note_catch_up_message_imported(&mut self) {
+		self.pending_catch_up_import.take();
+	}
+
+	fn handle_catch_up_request(
+		&mut self,
+		who: &PeerId,
+		request: CatchUpRequestMessage,
+		set_state: &environment::SharedVoterSetState<Block>,
+	) -> (Option<GossipMessage<Block>>, Action<Block::Hash>) {
+		if request.set_id != self.local_view.set_id {
+			return (None, Action::Discard(-1));
+		}
+
+		let last_completed_round = set_state.read().last_completed_round();
+		if last_completed_round.number < request.round.0 {
+			return (None, Action::Discard(-1));
+		}
+
+		trace!(target: "afg", "Replying to catch-up request for round {} from {} with round {}",
+			request.round.0,
+			who,
+			last_completed_round.number,
+		);
+
+		let mut prevotes = Vec::new();
+		let mut precommits = Vec::new();
+
+		for vote in last_completed_round.votes {
+			match vote.message {
+				grandpa::Message::Prevote(prevote) => {
+					prevotes.push(grandpa::SignedPrevote {
+						prevote,
+						signature: vote.signature,
+						id: vote.id,
+					});
+				},
+				grandpa::Message::Precommit(precommit) => {
+					precommits.push(grandpa::SignedPrecommit {
+						precommit,
+						signature: vote.signature,
+						id: vote.id,
+					});
+				},
+				_ => {},
+			}
+		}
+
+		let (base_hash, base_number) = last_completed_round.base;
+
+		let catch_up = CatchUp::<Block> {
+			round_number: last_completed_round.number,
+			prevotes,
+			precommits,
+			base_hash,
+			base_number,
+		};
+
+		let full_catch_up = GossipMessage::CatchUp::<Block>(FullCatchUpMessage {
+			set_id: request.set_id,
+			message: catch_up,
+		});
+
```

Comment:
Maybe we should update the view of the peer after sending the catch up message. In case of receiving the same catch-up request, do we respond again?

Reply:
Good point! I think we should add a validation here that the peer doesn't request a catch up for a round earlier than whatever his view is. If we update their view when we reply to a catch-up then this would prevent DoS but I'm not sure we should do it. I was thinking of adding a cache of last requests in a follow-up PR, and this cache would be used to prevent DoS.

Reply:
Updating the view here could be a bit weird, since they might have a neighbor packet in transit that is from before the view we expect them to have upon receiving the catch-up message.

I'd rather simply rate-limit the amount of catch-up messages a peer can send us.

Reply:
Well, that's probably not the right approach considering the PSM...what we should be doing is actually charging the peer a cost for every catch-up request. The hope being that if they spam us, they get disconnected fairly quickly.

Reply:
Added a static cost to answering a catch up request which is inverse of the benefit the other party gets.